kernel/src/mem.rs

```diff
@@ -178,7 +224,24 @@ impl ReadWrite for ReadWriteAppSlice {
                 // memory that the process has `allow`ed to the kernel, and will not permit
                 // the process to free any memory after it has been `allow`ed. This guarantees
                 // that the buffer is safe to convert into a slice here.
-                let slice = unsafe { slice::from_raw_parts_mut(self.ptr, self.len) };
+                //
```

Comment:
Question tangential to this PR:

How do we prevent a capsule from invoking `mut_map_or` recursively on the same `ReadWriteAppSlice` to get two `&mut` references to a single buffer? In other words: what happens if `fun` calls `mut_map_or` again? Will `process_map_or` not call the closure on the second invocation?

Reply:
I don't think we currently protect against this, which is pretty bad. I _think_ the usual thing to do in Rust would be to require a mutable borrow of the "parent" object here, but I'm not sure how that will collide with the rest of the ecosystem (i.e. expectations by capsule code regarding non-mutable reference to ReadWriteAppSlice). I'll analyze and create a PR accordingly, if you don't mind. Memory safety is hard. :)

Reply:
TakeCell style semantics might be easier to integrate

Reply:
It turns out that requiring a mutable borrow of the `ReadWriteAppSlice` itself only breaks two capsules (which I've just mechanically fixed so the borrow-checker is happy). Should be fixed in #2465. Thanks for reporting this @jrvanwhy!